pkg/frontend/querymiddleware/tenant_federation.go

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+
+	"github.com/grafana/dskit/tenant"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/querier/tenantfederation"
+)
+
+const (
+	tooManyTenantsTemplate = "too many tenant IDs present. max: %d actual: %d"
+)
+
+// federationMiddleware validates the number tenants in the provided context matches the
+// configuration for tenant federation and that the tenant IDs are legal.
+type federationMiddleware struct {
+	next Handler
+	cfg  tenantfederation.Config
+}
+
+func newFederationMiddleware(cfg tenantfederation.Config) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return &federationMiddleware{
+			next: next,
+			cfg:  cfg,
+		}
+	})
+}
+
+func (f *federationMiddleware) Do(ctx context.Context, r Request) (Response, error) {
+	ids, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeInternal, err.Error())
```

Comment:
what's the reason you went with TypeInternal instead?

Reply:
No real reason, just that my initial (dumb) thought was "err = 500". I'll change to `TypeBadData` since we don't want invalid user requests to cause 500s.